Vignette Implementation

VladVin opened the pull request:
Мой pull request

Commit:
Add 2 new files and modify photoeffects.hpp

valentina-kustikova commented:
Влад, надо выполнить слияние с текущим состоянием upstream. Смотрите обсуждение в этом pull request: https://github.com/UNN-VMK-Software/itlab-vision/pull/9.

VladVin commented:
Делаю, как в инструкции, при выполнении команды git merge upstream/master пишет Already up-to-date. Почему так? Ведь должны были появиться либо конфликты, либо сообщение об успехе, и как понимаю, в моем локальном мастере должны появиться слитые файлы.

druzhkov-paul commented:
Добрый день.

Конфликта на данном этапе быть не должно, т.к. вы работали не в мастере, а
в отдельной ветке. Конфликты будут, когда вы будете перемещать (rebase)
свою ветку.
Сравните состояние своей ветки master с состоянием upstream/master. После
git merge upstream/master они должны быть идентичны.

## 

Павел

22 марта 2014 г., 13:54 пользователь VladVin notifications@github.comнаписал:

> Делаю, как в инструкции, при выполнении команды git merge upstream/master
> пишет Already up-to-date. Почему так? Ведь должны были появиться либо
> конфликты, либо сообщение об успехе, и как понимаю, в моем локальном
> мастере должны появиться слитые файлы.
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/UNN-VMK-Software/itlab-vision/pull/12#issuecomment-38347403
> .

VladVin commented:
Добрый вечер. А как я могу сравнить состояние с веткой upstream/master? Могу только посмотреть состояние своей ветки master

valentina-kustikova commented:
Нет понятия upstream/master, есть upstream, а есть origin/master и ветка master, которая была получена с использованием fork из upstream. 

По идее ваша ветка master должна иметь то же состояние, что и upstream. Чтобы привести ее в актуальное состояние надо выполнить примерно следующую последовательность команд:
`
git checkout master
git pull
`
или вот такой вариант:
`
git merge upstream/master
git merge master
git checkout mybranch
`

VladVin commented:
Спасибо. Я делал это, пишет Already up-to-date. Значит все сделано?

valentina-kustikova commented:
Да, значит, что master имеет актуальное состояние. Но минуту назад я влила изменения из pull request Артема, поэтому проверяйте еще раз, должно отличаться.

VladVin commented:
Ага, я последние изменения сейчас себе сделал. Теперь выполняю команду и вижу
vlad@vlad-Q330:~/MyProjects/itlab-vision$ git merge upstream/master
error: Your local changes to the following files would be overwritten by merge:
    photoeffects/include/photoeffects.hpp
Please, commit your changes or stash them before you can merge.
Aborting

Он просит сделать коммит из master, но ведь мне нельзя этого делать?

valentina-kustikova commented:
Это означает, что внесли изменения в своей текущей ветке, но не закоммитили их на сервер в соответствующую ветку. А ветка master это или нет, это мне неизвестно :) Возможно, вы делали изменения не в своей ветке, а в ветке master, что не правильно. Проверьте командой `git branch`, какая у вас ветка активная (будет отмечена звездочкой). Потом командой `git status` состояние файлов в локальном репозитории. Там будут явно видны файлы, состояние которых не соответствует состоянию на сервере.

VladVin commented:
Такое состояние файлов в ветке master
`Changes to be committed:
(use "git reset HEAD <file>..." to unstage)
modified:   photoeffects/include/photoeffects.hpp
new file:   photoeffects/src/glow.cpp
modified:   photoeffects/src/sepia.cpp
new file:   photoeffects/src/tint.cpp
new file:   photoeffects/test/glow_test.cpp
new file:   photoeffects/test/tint_test.cpp`

Пытаюсь перейти в ветку Vignette_implementation, получаю
vlad@vlad-Q330:~/MyProjects/itlab-vision$ git checkout Vignette_implementation
error: Your local changes to the following files would be overwritten by checkout:
    photoeffects/include/photoeffects.hpp
Please, commit your changes or stash them before you can switch branches.
Aborting

и по-прежнему остаюсь в master:)

VladVin commented:
Странный размер текста, поменять не получается

valentina-kustikova commented:
Все правильно, это новые файлы, которые появились после слияния pull request, который делали Артем и Дмитрий.

Просто вы наделали изменений в ветке master, но на сервер в вашу ветку master ничего не закоммитили. Надо было сначала перейти в ветку Vignette_implementation, а потом вносить изменения и коммитить в ветку. После последнего коммита, когда вы убедились, что ничего не сломалось (тесты), переходить в master и мержить его с текущим состоянием ветки upsream.

Мое предложение - клонировать заново репозиторий, чтобы сохранить текущие изменения, а потом все сделать, как надо в смысле переходов по веткам и изменений.

ps. размер копируемого текста был обусловлен наличием символов решетка в начале строк, а это спец символ языка разметки Markdown.

VladVin commented:
Спасибо. А перед клонированием репозитория мне нужно удалить текущий, временно переместив куда-нибудь только созданные мною файлы?

valentina-kustikova commented:
Ничего удалять не надо, можно просто клонировать в другую директорию. Либо как вариант, можно куда-то сохранить изменения, а потом постепенно внести их в ветку.

VladVin commented:
Почитал еще книгу, сейчас все получается. Просит устранить конфликты в файле photoeffects/photoeffects.hpp. Какие изменения нужно сделать в файле, чтобы он принял его?

valentina-kustikova commented:
Влад, у вас опять неактуальные коды в вашей ветке master. Что конкретно вам мержить, я не знаю, надо видеть код :)

VladVin commented:
Сейчас у меня актуальные коды, такие же, как в центральном репозитории. Например, в папке src лежат 6 файлов, которые делали другие, плюс один мой vignette.cpp. Файл photoeffects/photoeffects.hpp содержит все прототипы функций других и мой. Там были стрелки <<<<<HEAD и прочее. я это удалил, оставив только прототипы. Теперь наверное нужно сделать add и коммит? Только я нахожусь ни в какой ветке после выполнения rebase. т.е. git branch пишет под звездочкой *(no branch)

valentina-kustikova commented:
rebase - это не ветка, а команда.

Влад, читайте дальше Pro Git.

VladVin commented:
Валентина, Вы меня неправильно поняли:) я как раз и имею в виду команду. После ее выполнения...

Commit:
changed vignette_test.cpp

valentina-kustikova commented on photoeffects/include/photoeffects.hpp:
```diff
@@ -15,3 +15,5 @@
 
 cv::Mat antique(cv::Mat& image);
+
+cv::Mat vignette(cv::Mat& image, float radius = 1);
```
В параметрах будет неявное приведение типов от целого к вещественному, если оставить 1. Поэтому надо явно указать, что это вещественное число.
`
float radius = 1.0f
`

VladVin replied:
Ага, спасибо. Но я все равно изменю этот прототип.

Commit:
modified 3 files

Commit:
Vignette implementation

valentina-kustikova commented on photoeffects/src/vignette.cpp:
```diff
@@ -0,0 +1,46 @@
+    Vec3b intensity, intensity_new;
+    double a = rect.height / 2.0f;
+    double b = rect.width / 2.0f;
+    double radius_ellipse = 0.0f;
```
А почему константы имеют тип `float`, а переменные - `double` (стр. 11-13, 15-18)? Думаю, что типа `float` для всех операций вполне достаточно.

valentina-kustikova commented on photoeffects/src/vignette.cpp:
```diff
@@ -0,0 +1,46 @@
+int vignette(InputArray src, OutputArray dst, Size rect)
+{
+    Mat img_src(src.size(), src.type());
+    img_src = src.getMat();
```
Надо проверить, что исходная матрица имеет нужный тип.

valentina-kustikova commented on photoeffects/src/vignette.cpp:
```diff
@@ -0,0 +1,46 @@
+    double a = rect.height / 2.0f;
+    double b = rect.width / 2.0f;
+    double radius_ellipse = 0.0f;
+    double max_radius = sqrt((img_src.rows / 2) * (img_src.rows / 2) + (img_src.cols / 2) * (img_src.cols / 2));
```
Если перейдете в типу `float`, то надо использовать `sqrtf`.

valentina-kustikova commented on photoeffects/src/vignette.cpp:
```diff
@@ -0,0 +1,46 @@
+            intensity = img_src.at<Vec3b>(i, j);
+            if ((i - img_src.rows / 2.0f) * (i - img_src.rows / 2.0f) / (a * a) +
+                    (j - img_src.cols / 2.0f) * (j - img_src.cols / 2.0f) / (b * b) > 1)
+            {
```
В циклах много инвариантов, которые можно вычислить один раз до циклов:
`
img_src.rows / 2.0f
img_src.cols / 2.0f
a * a
b * b
`

valentina-kustikova commented on photoeffects/src/vignette.cpp:
```diff
@@ -0,0 +1,46 @@
+        {
+            intensity = img_src.at<Vec3b>(i, j);
+            if ((i - img_src.rows / 2.0f) * (i - img_src.rows / 2.0f) / (a * a) +
+                    (j - img_src.cols / 2.0f) * (j - img_src.cols / 2.0f) / (b * b) > 1)
```
Откуда целочисленная константа `1`? Может, если исправим все в `float`, то `1.0f`?

valentina-kustikova commented on photoeffects/src/vignette.cpp:
```diff
@@ -0,0 +1,46 @@
+                dist = sqrt((i - img_src.rows / 2.0f) * (i - img_src.rows / 2.0f) +
+                            (j - img_src.cols / 2.0f) * (j - img_src.cols / 2.0f));
+                sin_fi = (img_src.rows / 2.0f - i) / dist;
+                radius_ellipse = (a * b) / sqrt(a * a * (1 - sin_fi * sin_fi) + b * b * sin_fi * sin_fi);
```
Если перейдете в типу `float`, то надо использовать `sqrtf`.

valentina-kustikova commented on photoeffects/src/vignette.cpp:
```diff
@@ -0,0 +1,46 @@
+            else
+            {
+                img_dst.at<Vec3b>(i, j) = intensity;
+            }
```
Можно обойтись без блока `else`, если перед условием сразу выполнить присваивание `img_dst.at<Vec3b>(i, j) = intensity;`

valentina-kustikova commented on photoeffects/test/vignette_test.cpp:
```diff
@@ -0,0 +1,10 @@
+{
+    //Mat image(10, 10, CV_8UC3);
+    //EXPECT_EQ(10, vignette(image, img_dst, 0, 0, 100));
+}
```
Нужны тесты конечно же.

valentina-kustikova commented:
Если будет время, то можно усложнить за счет переноса центра эллипса и его поворота (http://www.photoshopessentials.com/photo-effects/radial-filter-vignette-effect).

VladVin commented:
Спасибо за замечания, будет исправлено.

Commit:
modified vignette implementation

Commit:
Added warmify implementation and 2 files of rst-documentation

Commit:
Changed photoeffects.hpp and vignette.cpp

Commit:
deleted 3 files

Commit:
Modified vignette.cpp and added vignette_sample.cpp

Commit:
a few changes of vignette_sample.cpp

Commit:
Vignette.cpp was change by format code requirements

Commit:
changed google tests

VladVin commented:
Форматировал исходный код vignette.cpp по общим правилам, заменил if-else проверки на CV_Assert и изменил файл vignette_test.cpp

Commit:
vignette_sample.cpp was be changed

valentina-kustikova commented on photoeffects/samples/vignette_sample.cpp:
```diff
@@ -0,0 +1,38 @@
+    {
+        cout << "Error. Image wasn't found." << endl;
+        return -1;
+    }
```
Влад, посмотрите, как сделана обработка аргументов командной строки в фильтре сепии. Надо сделать по аналогии.

valentina-kustikova commented on photoeffects/samples/vignette_sample.cpp:
```diff
@@ -0,0 +1,38 @@
+    Mat img_dst;
+    Size rectangle;
+    rectangle.height = image.rows - 200;
+    rectangle.width = image.cols - 300;
```
А вы уверены, что после вычитания ширина и высота для `rectangle` будет положительной?

valentina-kustikova commented on photoeffects/samples/vignette_sample.cpp:
```diff
@@ -0,0 +1,38 @@
+    if (vignette(image, img_dst, rectangle) == 0)
+    {
+        imshow(argv[1], img_dst);
+        imwrite("/home/vlad/Изображения/girl2_vignette.jpg", img_dst);
```
Откуда такие пути?!

Если хочется сохранить изображение, что запросите полное имя результирующего изображения в качестве параметра командной строки.

valentina-kustikova commented on photoeffects/samples/vignette_sample.cpp:
```diff
@@ -0,0 +1,38 @@
+    rectangle.height = image.rows - 200;
+    rectangle.width = image.cols - 300;
+    imshow(argv[1], image);
+    if (vignette(image, img_dst, rectangle) == 0)
```
Не стоит писать условия с двумя ветками, когда можно без них обойтись. Надо написать что-то такое:
`
int opRes = vignette(image, img_dst, rectangle);
if (opRes == 1)
{
        cout << "Incorrect image type." << endl;
        return 2;
}
`

valentina-kustikova replied:
Дальнейший код конечно не надо в ветку `else` помещать.

valentina-kustikova commented on photoeffects/src/vignette.cpp:
```diff
@@ -0,0 +1,67 @@
+int vignette(InputArray src, OutputArray dst, Size rect)
+{
+    CV_Assert(src.type() == CV_8UC3 && rect.height != 0 && rect.width != 0);
+    Mat imgSrc(src.size(), CV_8UC3);
```
Можно не создавать матрицу, достаточно объявить переменную.

VladVin commented:
Спасибо за замечания.
1. Ага, исправлю.
2. Я сразу об этом подумал. Хотел сделать проверку, но попробовал присваивать отрицательные значения, они автоматически берутся положительными. Правда я не знаю почему. Будто там типы беззнаковые, хотя на самом деле double.
3. Пути я, конечно поменяю. Этот сэмпл был написан просто для вызова моей функции. Я же перепишу и сделаю с графическим интерфейсом.
4. Исправлю

valentina-kustikova commented on photoeffects/src/vignette.cpp:
```diff
@@ -0,0 +1,67 @@
+    int centerCol = imgSrc.cols / 2;
+    float aSquare = (rect.height / 2.0f) * (rect.height / 2.0f);
+    float bSquare = (rect.width / 2.0f) * (rect.width / 2.0f);
+    float ab = (rect.height / 2.0f) * (rect.width / 2.0f);
```
Стоит сразу посчитать, чему равно значение выражения в знаменателе произведения (стр. 30-32), имею ввиду `4.0f * 4.0f`.

valentina-kustikova commented on photoeffects/src/vignette.cpp:
```diff
@@ -0,0 +1,67 @@
+    float ab = (rect.height / 2.0f) * (rect.width / 2.0f);
+    float radiusEllipse = 0.0f;
+    float radiusMax = sqrt((imgSrc.rows / 2) * (imgSrc.rows / 2) +
+                           (imgSrc.cols / 2) * (imgSrc.cols / 2));
```
А это нормально, что везде целочисленное деление? Ведь результат тоже целое число, а не вещественное типа `float`.

valentina-kustikova commented on photoeffects/src/vignette.cpp:
```diff
@@ -0,0 +1,67 @@
+
+            intensityNew = intensity;
+            if ((i - centerRow) * (i - centerRow) / aSquare +
+                    (j - centerCol) * (j - centerCol) / bSquare > 1)
```
Вот здесь разности `i - centerRow` и `j - centerCol`, а также их квадраты можно посчитать один раз до условия, потому что внутри условного оператора они используются как в условии, так и внутри ветки.

valentina-kustikova commented on photoeffects/src/vignette.cpp:
```diff
@@ -0,0 +1,67 @@
+                    (j - centerCol) * (j - centerCol) / bSquare > 1)
+            {
+                dist = sqrtf((i - centerRow) * (i - centerRow) +
+                             (j - centerCol) * (j - centerCol));
```
Подкоренное выражение имеет целочисленный тип, а `sqrtf` работает с типом `float`. В некоторых компиляторах получите предупреждение, поэтому предлагаю здесь поставить явное преобразование типов данных, либо подумать над тем, нужен ли здесь вещественный корень.

valentina-kustikova commented on photoeffects/src/vignette.cpp:
```diff
@@ -0,0 +1,67 @@
+                             (j - centerCol) * (j - centerCol));
+                sinFi = (i - centerRow) / dist;
+                radiusEllipse = ab / sqrt(aSquare * (1 - sinFi * sinFi) +
+                                          bSquare * sinFi * sinFi);
```
Выражение `sinFi * sinFi` можно вычислить один раз.

valentina-kustikova replied:
Еще здесь функция корня `sqrt`, которая работает с `int` или `double`, хотя аргумент имеет тип `float`.

VladVin commented:
Я почему-то пропустил эти вещи. да, конечно, исправлю.

valentina-kustikova commented on photoeffects/src/vignette.cpp:
```diff
@@ -0,0 +1,67 @@
+
+                intensityNew.val[0] = (uchar)(intensity.val[0] * coefficient);
+                intensityNew.val[1] = (uchar)(intensity.val[1] * coefficient);
+                intensityNew.val[2] = (uchar)(intensity.val[2] * coefficient);
```
Думается мне, что для класса `Vec3b` перегружена операция умножения на скаляр. А если и не перегружено умножение, то точно перегружена операция `[]`.

valentina-kustikova commented on photoeffects/test/vignette_test.cpp:
```diff
@@ -0,0 +1,13 @@
+    rectangle.width = image.cols / 2;
+    rectangle.height = image.rows / 2;
+    EXPECT_EQ(0, vignette(image, dst, rectangle));
+}
```
Посмотрите тесты для фильтра сепии. Этого недостаточно.

VladVin commented:
Хорошо

Commit:
implementation of vignette and vignette_sample was be increased

VladVin commented:
Применил все замечания по коду и улучшил vignette. в сэмпле также учел замечания.

valentina-kustikova commented on photoeffects/samples/vignette_sample.cpp:
```diff
@@ -0,0 +1,49 @@
+using namespace std;
+
+const char *helper =
+"./sepia_sample <img>\n\
```
Откуда здесь название бинарника `./sepia_sample`? У вас бинарник явно будет по-другому называться.

VladVin replied:
это я скопировал из файла сепии и не обратил внимания. сейчас изменю:)

valentina-kustikova commented on photoeffects/samples/vignette_sample.cpp:
```diff
@@ -0,0 +1,49 @@
+        return 1;
+    }
+
+    image = imread(argv[1], CV_LOAD_IMAGE_COLOR);
```
Обработка аргументов командной строки так и не сделана.

VladVin replied:
Я хотел по-своему сделать. Ладно, если так будет правильней, то сейчас изменю.

valentina-kustikova commented on photoeffects/samples/vignette_sample.cpp:
```diff
@@ -0,0 +1,49 @@
+    imshow(dstImgWinName, vignetteImg);
+    waitKey();
+    destroyAllWindows();
+	return 0;
```
Отступ.

valentina-kustikova commented on photoeffects/src/vignette.cpp:
```diff
@@ -0,0 +1,70 @@
+// dst - the image with vignette effect
+// RESULT
+// Error status
+*/
```
Можно убрать.

valentina-kustikova commented on photoeffects/src/docs/vignette.rst:
```diff
@@ -0,0 +1,28 @@
+			**coefficient = 1 - ((dist - radius_ellipse) / (max_radius - radius_ellipse))**,
+			where max_radius is a distance between the pixel (0, img_src.cols) and the center image.
+
+#. Save this matrix as BGR image.
```
Нужны картинки с исходным изображением и результирующим.

Commit:
Processing arguments in sample was be added

Commit:
Image added to rst-documentation

druzhkov-paul commented:
@VladVin , посмотрите, что изменилось в плане форматирования в вашей документации, потому что у вас было местами неправильное оформление. Попробуйте собрать свою версию в html и pdf на своей машине.

Для ваших документаций пришлось создать новые ветки и перенести туда файлы. Ссылку на вас поставил в соответствующих commit'ах, pull request сделал от своего имени. Соответственно из ваших веток документацию теперь надо убрать.

valentina-kustikova commented on photoeffects/samples/vignette_sample.cpp:
```diff
@@ -0,0 +1,59 @@
+    {
+        cout << "Error. Image wasn't found." << endl;
+        return 2;
+    }
```
Это должно быть проверено через `CV_Assert` в вашем фильтре. Поэтому можно безболезненно убрать.

valentina-kustikova commented on photoeffects/src/vignette.cpp:
```diff
@@ -0,0 +1,70 @@
+    float dist = 0.0f;
+    float sinFi = 0.0f;
+    float sinFiSquare = 0.0f;
+    float coefficient = 0.0f;
```
Объявление переменных из строк 33, 35-41 можно перенести в циклы (туда, где они вычисляются), нет смысла их создавать здесь и инициализировать нулями.

valentina-kustikova commented on photoeffects/src/vignette.cpp:
```diff
@@ -0,0 +1,70 @@
+
+            jMinusCenterColSquare = (j - centerCol) * (j - centerCol);
+            intensityNew = intensity;
+            if (iMinusCenterRowSquare / aSquare + jMinusCenterColSquare / bSquare > 1)
```
Константа `1.0f`.

valentina-kustikova commented on photoeffects/src/vignette.cpp:
```diff
@@ -0,0 +1,70 @@
+            imgDst.at<Vec3b>(i, j) = intensityNew;
+        }
+    }
+    imgDst.convertTo(dst, CV_8UC3);
```
Зачем эта конвертация, если типы матриц совпадают?! Сразу можно работать с результирующей матрицей.

valentina-kustikova commented on photoeffects/test/vignette_test.cpp:
```diff
@@ -0,0 +1,13 @@
+    Mat image(100, 100, CV_8UC3), dst;
+    Size rectangle;
+    rectangle.width = image.cols / 2;
+    rectangle.height = image.rows / 2;
```
А в сэмпле у вас вещественное деление. Чему верить?

Commit:
Sample and implementation was be changed by according notes

valentina-kustikova commented on photoeffects/samples/vignette_sample.cpp:
```diff
@@ -0,0 +1,54 @@
+    {
+        cout << "Incorrect image type or size of rectangle." << endl;
+        return 2;
+    }
```
Не актуальна обработка кодов ошибки с `CV_Assert`. Используйте обработку исключений.

valentina-kustikova commented:
Надо обновить состояние ветки master.

valentina-kustikova commented on photoeffects/src/vignette.cpp:
```diff
@@ -0,0 +1,48 @@
+        float iMinusCenterRowSquare = iMinusCenterRow * iMinusCenterRow;
+        for (int j = 0; j < imgSrc.cols; j++)
+        {
+            intensity = imgSrc.at<Vec3b>(i, j);
```
Переменную `intensity` можно здесь же и объявить.

valentina-kustikova commented on photoeffects/src/vignette.cpp:
```diff
@@ -0,0 +1,48 @@
+            intensity = imgSrc.at<Vec3b>(i, j);
+
+            float jMinusCenterColSquare = (j - centerCol) * (j - centerCol);
+            intensityNew = intensity;
```
Переменную `intensityNew` можно здесь же объявить.

druzhkov-paul replied:
Ее вообще можно нигде объявлять. =) Зачем она вам нужна?

valentina-kustikova replied:
Могу мотивировать это только преждевременной оптимизацией, поэтому и не стала писать, что надо это удалить.

valentina-kustikova commented on photoeffects/src/vignette.cpp:
```diff
@@ -0,0 +1,48 @@
+                float radiusEllipse = ab / sqrtf(aSquare * (1.0f - sinFiSquare) +
+                                          bSquare * sinFiSquare);
+                float coefficient = 1.0f - ((dist - radiusEllipse) /
+                                      (radiusMax - radiusEllipse));
```
Лишние скобки для всего отношения.

Commit:
Documentation was be deleted

Commit:
vignette_sample was corrected by according notes

Commit:
tests was be corrected

Commit:
VignetteInavalidArguments in vignette_test.cpp  was be corrected

druzhkov-paul commented on photoeffects/src/docs/vignette.rst:
```diff
@@ -1,7 +1,6 @@
 =========================================
-Vignette
+vignette
```
Оставьте, пожалуйста, заголовок с большой буквы, как везде.

druzhkov-paul commented on photoeffects/src/docs/warmify.rst:
```diff
@@ -1,23 +1,21 @@
 =========================================
-Warmify
+warmify
```
Ну а этот-то файл как вообще сюда попал? И не меняйте в нем ничего, пожалуйста. Если вам что-то принципиально не нравится в текущей документации, можете сделать отдельный pull request, где предложите свой вариант.

druzhkov-paul commented on photoeffects/test/vignette_test.cpp:
```diff
@@ -0,0 +1,27 @@
+    rectangle.width = image.cols / 2.0f;
+
+    EXPECT_EQ(0, vignette(image, dst, rectangle));
+}
```
Тесты, нужны регрессионные тесты, которые проверят, то ваша функция не только отрабатывает без ошибки, но и выдает нужный результат.
